Stable order of platform and device objects

gmlueck opened the pull request:
Clarify that the order of the `platform` and `device` objects is the
same each time `get_platforms()` or `get_devices()` is called.  This
allows an application to store the index of a platform or device and
use that index later to identify the same platform or device.

Also clarify that the `get_devices()` function returns only the root
devices, not any sub-devices that have been created.

Closes internal issue 603.

tomdeakin commented:
I think we need to define "root device", given that we use the words parent and descendant when referring to devices and sub-devices. Also, it's possible there is an interpretation for root device conflating with a [PCIe root complex device](https://en.wikipedia.org/wiki/Root_complex).

Note the SYCL platform model section doesn't talk about sub-devices at all, which might complicate the issue.

It might be simplest to just add a glossary entry for "root device" and add cross-references/update the phrasing in any place we mean a parent device and not a sub-device.

What do you think?

gmlueck commented:
> I think we need to define "root device"

Yes, that's a good point.  I added a glossary entry for "root device" and changed the words "root device" to be a link to this glossary entry.  This was done in 5e101eeef8482b97ed00dba1ef3d56b36717d145

tomdeakin reviewed (approved):
LGTM now

fraggamuffin commented:
LGTM
